funding/manager.go

```diff
@@ -3601,3 +3659,144 @@ func (f *Manager) deleteChannelOpeningState(chanPoint *wire.OutPoint) error {
 	}, func() {})
 }
+
+// Here we wait for confirmation to update the real short channel id for this
+// not yet confirmed channel.
+func (f *Manager) updateShortChannelID(ch *channeldb.OpenChannel) (
+	*lnwire.ShortChannelID, error) {
+
+	log.Infof("waiting for confirmation to update short channel id")
+	confirmedChan := ch
+	confChan := make(chan *confirmedChannel, 1)
+	cancelChan := make(chan struct{})
+	skipChan := make(chan lnwire.ShortChannelID)
+
+	f.wg.Add(1)
+	go f.waitForFundingConfirmation(ch, cancelChan, confChan, 1)
+	defer close(skipChan)
+	defer close(cancelChan)
+
+	select {
+	case conf, ok := <-confChan:
+		var err error
+		if !ok {
+			return nil, fmt.Errorf("failed to wait for funding confirmation")
+		}
+		confirmedChan, err = f.cfg.FindChannel(lnwire.NewChanIDFromOutPoint(&ch.FundingOutpoint))
+		if err != nil {
+			return nil, fmt.Errorf("failed to find channel %v %v", ch.ShortChannelID, err)
+		}
+		err = f.cfg.Wallet.ValidateChannel(confirmedChan, conf.fundingTx)
+		if err != nil {
+			return nil, fmt.Errorf("unable to validate channel: %v", err)
+		}
+		if err := confirmedChan.MarkAsOpen(conf.shortChanID); err != nil {
+			return nil, fmt.Errorf("error setting channel pending flag to "+
+				"false: %v", err)
+		}
+		if err := f.cfg.ReportShortChanID(confirmedChan.FundingOutpoint); err != nil {
+			log.Errorf("unable to report short chan id: %v", err)
+		}
+	case <-f.quit:
+		// The fundingManager is shutting down, and will resume wait on
+		// startup.
+		return nil, ErrFundingManagerShuttingDown
+	}
+	return &confirmedChan.ShortChannelID, nil
+}
+
+type fakeIDsManager struct {
+	db *channeldb.DB
+}
+
+func newFakeIDsManager(channels []*channeldb.OpenChannel,
+	graph *channeldb.ChannelGraph, db *channeldb.DB) (*fakeIDsManager, error) {
+
+	// We would like to prune from the fake ids bucket all those that are not
+	// in use anymore. We start by create a mapping between ids to channels.
+	idsToChannels := make(map[uint64]*channeldb.OpenChannel, len(channels))
+	for _, c := range channels {
+		shortID := c.ShortChannelID
+		if shortID.IsFake() {
+			idsToChannels[shortID.ToUint64()] = c
+		}
+	}
+
+	// We marke each fake id in our bucket that is not associated with
+	// a pending or opened channel for prune.
+	var idsToPrune [][]byte
+	err := kvdb.Update(db, func(tx kvdb.RwTx) error {
+		fakeIDsBucket, err := tx.CreateTopLevelBucket(fakeIDsBucket)
+		if err != nil {
+			return err
+		}
+		return fakeIDsBucket.ForEach(func(k, v []byte) error {
+			shortID := byteOrder.Uint64(k)
+			_, exists := idsToChannels[shortID]
+			if !exists {
+				log.Infof("prunning fake id from graph %v", shortID)
+				idsToPrune = append(idsToPrune, k)
+			}
+			return nil
+		})
+	}, func() {})
+	if err != nil {
+		return nil, err
+	}
+
+	// We prune first from the graph.
+	for _, id := range idsToPrune {
+		shortID := byteOrder.Uint64(id)
+		if err = graph.DeleteChannelEdges(false, shortID); err != nil {
```

Comment:
Why not batch this call? Mainly to be able to distinguish which IDs have already been pruned?

Reply:
> Why not batch this call? Mainly to be able to distinguish which IDs have already been pruned?

Mainly because the DeleteChannelEdges exit with error on first edge that is not found and we may end up with edges that are not pruned due to this call keep returning error.